### PR TITLE
WIP: ENH: cython port of ctraits.c

### PR DESCRIPTION
**Pull request by dpinte**
This PR is the result of work started in 2013 to migrate `ctraits.c` to Cython. Thanks to the input of multiple developers, it comes in as a clean enhancement over the current code base, a test suite that passe without any changes. It open the doors to a many future enhancements.

Thanks to @jvkersch, @pberkes and @sjagoe for the work!

### CODE REVIEW

**Comment by pberkes** [comment]
This is great, thanks!

I'm running the test suite of a recent project using this branch, and I'm getting this error a lot:
```
  [...]
  File "traits/ctraits.pyx", line 886, in traits.ctraits.CHasTraits.__getattr__ (traits/ctraits.c:10109)
    value = trait.getattr(trait, self, name)
  File "traits/ctraits.pyx", line 1370, in traits.ctraits.getattr_constant (traits/ctraits.c:16755)
    raise NotImplementedError('getattr constant NOT IMPL.')
NotImplementedError: getattr constant NOT IMPL.
```

I had to remove the first part of the traceback to keep some information confidential, I'll try to create a minimal example.

**Comment by pberkes** [comment]
Here's the example:
```
from traits.api import *
class A(HasTraits):
    txt = Constant('something')
a = A()
print a.txt
```

This gives:
```
Traceback (most recent call last):
  File "/Users/pberkes/del/crash.py", line 5, in <module>
    print a.txt
  File "traits/ctraits.pyx", line 886, in traits.ctraits.CHasTraits.__getattr__ (traits/ctraits.c:10109)
    value = trait.getattr(trait, self, name)
  File "traits/ctraits.pyx", line 1370, in traits.ctraits.getattr_constant (traits/ctraits.c:16755)
    raise NotImplementedError('getattr constant NOT IMPL.')
NotImplementedError: getattr constant NOT IMPL.
```

**Comment by mdickinson** [comment]
Marking this PR as WIP. It sounds like it's still quite a long way from full compatibility with Traits.

**Comment by mdickinson** [comment]
See #301 for additional comments about test failures.

**Comment by sjagoe** [comment]
This appears to have lost all of the `PyUnicode_Check()` calls that are hidden behind defines (e.g. `Py2to3_SimpleString_Check()` in `ctraits.c`

**Comment by jvkersch** [comment]
@sjagoe I don't know much about the genesis of this branch but I thought those checks were introduced with the port to Python 3; doesn't this branch pre-date that work?

**Comment by sjagoe** [comment]
Yes, probably. We just need to keep in mind that we will need to update this to be master-compatible in addition to fixing the other bugs.

**Comment by dpinte** [comment]
@jvkersch the first commits are from Feb 2013. Not sure when the Py3 port arrived but IIRC it was after ...

File: traits/tests/test_disallow.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+class TestDisallow(unittest.TestCase):
+
+    def test_trait_get_set(self):
+        a = A()
+
+        with self.assertRaises(AttributeError):
+            a.x
+
+        with self.assertRaises(TraitError):
+            a.x = 123
```
Comment (sjagoe) [inline traits/tests/test_disallow.py]: Should this test assert that `a.other_attribute = 'something'` is allowed?

**Comment by dpinte** [comment]
Down to 3 errors:

```
======================================================================
ERROR: test_delegate_event_raises_attribute_error (traits.tests.test_delegate_event.TestDelegatesError)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dpinte/projects/traits/traits/tests/test_delegate_event.py", line 19, in test_delegate_event_raises_attribute_error
    c.button
  File "traits/ctraits.pyx", line 912, in traits.ctraits.CHasTraits.__getattr__ (traits/ctraits.c:11174)
    raise ValueError('getattr cannot be null ...')
  File "traits/ctraits.pyx", line 1384, in traits.ctraits.getattr_delegate (traits/ctraits.c:17983)
DelegationError: The '<class 'traits.tests.test_delegate_event.Child'>' object has no attribute 'button' because its <class 'traits.tests.test_delegate_event.Parent'> delegate has no attribute 'button'.

======================================================================
ERROR: test_this_none (traits.tests.test_traits.TestThis)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dpinte/projects/traits/traits/tests/test_traits.py", line 975, in test_this_none
    d.allows_none = None
  File "traits/ctraits.pyx", line 926, in traits.ctraits.CHasTraits.__setattr__ (traits/ctraits.c:11359)
    def __setattr__(self, name, value):
  File "traits/ctraits.pyx", line 954, in traits.ctraits.CHasTraits._internal_setattr (traits/ctraits.c:11614)
    else:
  File "traits/ctraits.pyx", line 1494, in traits.ctraits.setattr_trait (traits/ctraits.c:19181)
    if not PyString_Check(name):
  File "traits/ctraits.pyx", line 240, in traits.ctraits.validate_trait_self_type (traits/ctraits.c:3495)
    if (len(trait.py_validate) == 2 and value is None) or isinstance(value, type(obj)):
NotImplementedError: vt

======================================================================
ERROR: test_this_other_class (traits.tests.test_traits.TestThis)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dpinte/projects/traits/traits/tests/test_traits.py", line 992, in test_this_other_class
    d.allows_none = object()
  File "traits/ctraits.pyx", line 926, in traits.ctraits.CHasTraits.__setattr__ (traits/ctraits.c:11359)
    def __setattr__(self, name, value):
  File "traits/ctraits.pyx", line 954, in traits.ctraits.CHasTraits._internal_setattr (traits/ctraits.c:11614)
    else:
  File "traits/ctraits.pyx", line 1494, in traits.ctraits.setattr_trait (traits/ctraits.c:19181)
    if not PyString_Check(name):
  File "traits/ctraits.pyx", line 240, in traits.ctraits.validate_trait_self_type (traits/ctraits.c:3495)
    if (len(trait.py_validate) == 2 and value is None) or isinstance(value, type(obj)):
NotImplementedError: vt

----------------------------------------------------------------------
Ran 556 tests in 29.756s

FAILED (SKIP=8, errors=3)
```

**Comment by mdickinson** [comment]
This was an interesting project, but at this point the PR branch has diverged so much from master that I think it's not useful to keep this PR open. Replacing ctraits.c with a Cython version may still be something we want to consider, but I think we'll want to revisit our motivations before doing so (and possibly write an EEP!).

I'm going to close this PR and delete the branch, but tag the final commit as "experimental/ctraits-in-cython" for ease of retrieval.

**Comment by mdickinson** [comment]
Tagged here: https://github.com/enthought/traits/releases/tag/experimental%2Fcythonize-ctraits

**Pull request closed without merging**